code/controllers/subsystem/pathfinder.dm

```diff
@@ -1,48 +1,664 @@
+	return ..()
+
+/**
+  * Warns and logs when a pathfinding operation times out. Since we don't want to add more overhead, we can't terminate it early, but we can record it.
+  */
+/datum/controller/subsystem/pathfinder/proc/warn_overtime(message)
+	message_admins("Pathfinding Timeout: [message]")
+	CRASH(message)
+
+///////////////////////////////////////////////////
+//////// COMMON ASTAR CODE START //////////////////
+///////////////////////////////////////////////////
+
+#define MANHATTAN(A, B)		(abs(A.x - B.x) + abs(A.y - B.y))
+#define BYOND(A, B)		get_dist(A, B)
+#define EUCLIDEAN(A, B)		(sqrt((A.x - B.x)**2 + (A.y - B.y)**2))
+
+// Let's sing the listmos song!
+
+/// Total size of node "list", which is indexed by turf.
+#define NODE_LIST_LENGTH 7
+/// Since we're using turfs, this is just the previous turf.
+#define NODE_PREVIOUS 1
+/// AStar node weight, lower is better, the list will be sorted by this
+#define NODE_WEIGHT 2
+/// Node cost from root
+#define NODE_COST 3
+/// Node heuristic cost to target
+#define NODE_HEURISTIC 4
+/// Node depth from root
+#define NODE_DEPTH 5
+/// Direction to expand in
+#define NODE_DIR 6
+/// Node's actual turf because I forgot to do associative list and I'm not sure assoc would even perform better
+#define NODE_TURF 7
+
+
+// forgive me, for this (and everything below it) is a sin.
+#ifdef PATHFINDING_DEBUG
+	#define SETUP_NODE(list, previous, cost, heuristic, depth, dir, turf) \
+		__INJECTING_NODE = NODE(previous, cost, heuristic, depth, dir, turf); \
+		if(visualize_pathfinding) { \
+			current_node_effect = new(turf); \
+			debug_effects += current_node_effect; \
+			current_node_effect.appearance = debug_appearance_node; \
+			debug_turf_to_node[turf] = current_node_effect; \
+			current_node_effect.color = node_color_potential; \
+			current_node_effect.alpha = debug_visual_alpha; \
+			node_by_turf[turf] = __INJECTING_NODE; \
+		}; \
+		INJECT_NODE(list, __INJECTING_NODE)
+#else
+	#define SETUP_NODE(list, previous, cost, heuristic, depth, dir, turf) \
+		__INJECTING_NODE = NODE(previous, cost, heuristic, depth, dir, turf); \
+		node_by_turf[turf] = __INJECTING_NODE; \
+		INJECT_NODE(list, __INJECTING_NODE)
+#endif
+
+/// Sets up variables needed for binary insert to avoid variable def overhead
+#define INJECTION_SETUP \
+	var/__INJECTION_LISTLEN; \
+	var/__INJECTION_LEFT; \
+	var/__INJECTION_RIGHT; \
+	var/__INJECTION_MID; \
+	var/list/__INJECTING_NODE;
+
+/// Binary inserts a node into the node list. Snowflake implementation to avoid overhead, as this will never handle datums. Lowest weighted nodes go towards the end of the list as list.len-- is faster when popping.
+#define INJECT_NODE(heaplist, nodelist) \
+	__INJECTION_LISTLEN = length(heaplist); \
+	if(!__INJECTION_LISTLEN) { \
+		heaplist[++heaplist.len] = nodelist; \
+	}; \
+	else { \
+		__INJECTION_LEFT = 1; \
+		__INJECTION_RIGHT = __INJECTION_LISTLEN; \
+		__INJECTION_MID = (__INJECTION_LEFT + __INJECTION_RIGHT) >> 1;\
```

Comment:
since `__INJECTION_LEFT = 1` is always true
this should just be `__INJECTION_RIGHT >> 1` shouldn't it

Reply:
should probably also add a comment this is just equal to integer div / 2 of the value